Add option to launch updater from sdw-notify script

rocodes opened the pull request:
## Status

Ready for review

## Description of Changes

Fixes #595 

Move Notify dialog into separate file and allow user to launch Preflight updater directly from the Notify dialog. 

Also separates out constants from Notify.py and Updater.py into a constants file (`sdw_util/Const.py`). Towards goals of a) reducing mocking of class or module internals during testing, b) reducing dependency surface/moving towards separation of UI and logic components. I've refactored Notify.py but not Updater.py to use the constants file. If this approach works for people then in a subsequent PR I'll also make some similar changes to Updater.py. 

Misc: 
- I've stuck to the [edit for clarity] camel case naming convention going on here (NotifyApp.py), but I'd be happy to switch to more typical Python naming conventions, possibly in another PR so as to keep things clean. 

- As discussed in sdw developer hangout, a separate PR will be made to begin adding UI tests for all the launcher/notifier components.

### Changes proposed in this pull request:

- Replace simple notification ("ok") with option to launch updater from sdw-notify script, or to defer reminder (deferral period is until the next time the cron job runs)
- Change cron job that runs notifier to run every 2 hours instead of hourly  

## Testing

Check out this branch in your sd dev vm.

**Manual testing**
Copy this branch to dom0 via `make clone`, then `make dev`. 

Ensure `sd-app` VM is shut down. 
Edit the file in `dom0` called `~/.securedrop_launcher/sdw-last-updated` to change the date to a date >= 5 days ago.
Run `python3 /opt/securedrop/launcher/sdw-notify.py`:
- [ ] Notification appears with prompt to check for updates. 
- [ ] No "**Warning:**" text is present in the notification
- [ ] The default selected button option is "Check for updates".
- [ ] Clicking "Remind me later" (or pressing the Esc key) closes the dialog window. Note the time.

Start `sd-app` vm via `qvm-start sd-app` in `dom0`. Wait for the notify prompt to reappear.  
- [ ] The notify prompt appears again after a delay of t < 2 hours (The reminder runs every 2 hours on the even hours. The delay will depend on what time the user dismissed the dialog). 
- [ ] This time, the "**Warning:**" text is present in the notification
- [ ] Clicking "Check for updates" launches the Preflight Updater.  
- [ ] The Preflight Updater can run updates successfully. 
- [ ] No GTK warnings or other errors are generated or printed to the `dom0` console when the Preflight Updater is launched.

**Automated testing**
- [ ] All tests pass

## Deployment

Any special considerations for deployment? Consider both:

1. Upgrading existing pilot instances
2. New installs

## Checklist

### If you have made changes to the provisioning logic

- [ ] All tests (`make test`) pass in `dom0`

### If you have added or removed files

- [ ] I have updated `MANIFEST.in` and `rpm-build/SPECS/securedrop-workstation-dom0-config.spec`

### If documentation is required

- [ ] I have opened a PR in the [docs repo](https://github.com/freedomofpress/securedrop-workstation-docs) for these changes, or will do so later
- [ ] I would appreciate help with the documentation

gonzalo-bulnes commented on launcher/sdw-notify.py:
```diff
@@ -54,17 +54,44 @@ def show_update_warning():
-        QMessageBox.Ok,
-        QMessageBox.Ok,
-    )
+    dialog = QMessageBox()
+    dialog.setWindowTitle(strings.headline_notify_updates)
+    dialog.setIcon(QMessageBox.Warning)
+    dialog.setStandardButtons(QMessageBox.No | QMessageBox.Ok)
+    dialog.setDefaultButton(QMessageBox.Ok)
+    button_check_now = dialog.button(QMessageBox.Ok)
+    button_check_now.setText(strings.button_check_for_updates)
+    button_defer = dialog.button(QMessageBox.No)
+    button_defer.setText(strings.button_defer_check)
+
+    if Util.is_sdapp_halted():
+        dialog.setText(strings.description_notify_updates)
+    else:
+        # If sd-app is running or state is unknown, use longer warning message
+        dialog.setText(strings.description_notify_updates_sdapp_running)
+
```
I think this dialog could be extracted as a standalone component (widget).
The way I see it, it is a static dialog (the contents wouldn't change while the dialog is open) so the condition `Util.is_sdapp_halted()` could be a parameter to the constructor.

The API of the new class would be the same as the `QMessageBox` API, so none of the code below should need any change, that's what makes me think the refactoring could be a good idea.

gonzalo-bulnes replied:
Naming-wise, my first thoughts would go towards "update prompt", "security check prompt"... what do you think @rocodes?

gonzalo-bulnes commented on launcher/sdw-notify.py:
```diff
@@ -4,11 +4,10 @@
 
-from sdw_notify import Notify
-from sdw_updater_gui import Updater
+from sdw_notify import Notify, strings
```
I really like the separate module of strings.

If we were to extract the dialog below, then those strings could be encapsulated within that namespace :slightly_smiling_face: 

Not sure if that's the most idiomatic approach, but when thinking of reducing the API surface, I think:

```sh
tree .
 .
└── security_check         # or whatever name makes sense, it's the one that will show up in imports
           ├── __init__    # exposes the dialog class
           ├── prompt.py   # or whatever is a good name for the dialog class
           └── strings.py  # imported by the file above but not exposed by __init__
``` 

Usage could look like (just examples, I picked names to illustrate what I'd try to expose from the module):

```py
import security_check
user_choice = security_check.Prompt(sdapp_halted=True).exec()
# I personally like this best, because there is no namespace collapse, "Prompt" is always namspaced.
```

or

```py
from security_check import SecurityCheckPrompt
user_choice = SecurityCheckPrompt(sdapp_halted=True).exec()
# Might be more idiomatic? I don't like too much that "SecurityCheckPrompt" suddenly is exported by this module too - :grimacing: 
```

rocodes replied:
@gonzalo-bulnes see if the changes in ed8d97f work for you in terms of separation of concerns :)

I _was_ hoping that I could completely eliminate the dependency on Qt components in `sdw-notify.py`, which I think would be very clean, but that requires just a little more time and thought to avoid issues with creating/handling `QApplication` (managing event loops). My goal would be to start here, perhaps make some similar improvements to the `sdw-launcher` in terms of improving the way it does its imports, and eventually move away from dependencies on Qt components in all but the view components. But I'd like to hear your thoughts.

gonzalo-bulnes reviewed (commented):
@rocodes As discussed earlier, here are a few thoughts re: **opportunity to extract the UI / dialog** out of the notifier. I've given temporary names to things, but it's really jsut to illustrate the general idea, see if you think it's something you'd want to do.


About **testing the UI**: I too have been thinking about that (mostly with https://github.com/freedomofpress/securedrop-client/issues/1345 in mind). This is what I've got to so far: [example](https://github.com/freedomofpress/securedrop-client/pull/1341/files#diff-b4d9842eb21d46ff58d89e9b4453eac3c37b21f2b9b586d36aeb31cb6e8b1aa2), [another example](https://github.com/freedomofpress/securedrop-client/pull/1346/files#diff-30ac31cdba573941cb1b5b9c12f5414f995538f23391b300d20aeb84bdb52f11) with [comments](https://github.com/freedomofpress/securedrop-client/pull/1346#discussion_r754943522).

rocodes commented:
`make black` passes on my local branch (for sanity, `make black --check --line-length=100` reports no changes would be made). Just trying to see what the difference is.

gonzalo-bulnes commented on launcher/tests/test_util.py:
```diff
@@ -302,3 +302,56 @@ def test_get_logger():
     assert logger.name == "{}.{}".format(test_prefix, test_module)
+
+
+def test_is_sdapp_halted_runs_once():
+    """
+    When check for sd-app state is run
+     Then exactly one subprocess call should be run
+    """
+    with mock.patch("subprocess.check_output") as patched_subprocess_check:
+        util.is_sdapp_halted()
+        patched_subprocess_check.assert_called_once()
+
+
```
This sub-process call doesn't have any side-effects, it is used _for the value it returns_.

From that point of view, I don't think it is needed or useful to assert that it is called, as long as the behavior of the function (`util.is_sdapp_halted`) is tested for all the possible return values of the sub-process call. And that's what you did with the two tests below. :+1: 

On the other size, these call assertions are a bit brittle: they tend to break if the number of calls or the parameters change, in ways that often don't reflect an actual change of behavior of the function under test ("false positives/negatives"), and get in the way of refactoring.

Because of that I think it's safe, and I'd suggest removing this test.

gonzalo-bulnes reviewed (dismissed):
I think this does a great job at separating concerns! :heavy_check_mark: The `QApplication` event loop is indeed a bit tricky to extract, we can think of ways of doing that (I'll have the same question for the Client).

I'd remove one of the tests (specially because I've encountered a bunch of failing-for-no-good-reason tests yesterday so the feeling is fresh :grimacing:), and I left you a couple of naming thoughts, but none of these are blockers, so fell free to proceed! :slightly_smiling_face:

eloquence commented on launcher/sdw_notify/Notify.py:
```diff
@@ -16,19 +16,6 @@
 LAST_UPDATED_FILE = os.path.join(BASE_DIRECTORY, "sdw-last-updated")
 LAST_UPDATED_FORMAT = "%Y-%m-%d %H:%M:%S"
 
-# The lockfile basename used to ensure this script can only be executed once.
-# Default path for lockfiles is specified in sdw_util
-LOCK_FILE = "sdw-notify.lock"
```
In truth, package users should disregard these variables - they're intended for developer use, defined as shared pseudo-constants to make them easier to modify in one place. If we ever expose any aspect of the updater to user configuration, it would likely be via the configuration file `config.json` in `/usr/share/securedrop-workstation-dom0-config` in `dom0`.

gonzalo-bulnes replied:
By "package users" I meant devs, not end-users! Sorry my wording was unclear. In my mind the "package" being a Python concept is used by devs. Anyhow, I agree in not thinking there is much/anything we'd want end-users (SecureDrop admins?) to configure in the foreseeable future. (The workstation is probably complex enough with reasonably set defaults!)

rocodes replied:
This is not an essential part of the changes, and is definitely the part I thought might be contentious ;) I'm open to reverting it, and/or changing the name of the file if that's the sticking point. (Condensed rationale for splitting out the variables into a separate file is in the PR description)

gonzalo-bulnes replied:
Certainly **not** a sticking point!

Clarification: I was trying to find out what package name would best describe what constants are expected in it, because I think the rationale for splitting them apart makes sense and I anticipate we're likely to want to repeat that in the future. At that point, the question will likely be: do this constant and that one belong together or should they be in distinct files? At that point I feel that `Const` as a name doesn't help much making distinctions.

Let's cross that naming bridge when we get there!

eloquence commented:
Re: caps for constants
> I'd be happy to switch to more typical Python naming conventions, 

AFAIK the all-caps style is idiomatic, see https://www.python.org/dev/peps/pep-0008/#constants

rocodes commented on launcher/sdw_notify/NotifyApp.py:
```diff
@@ -0,0 +1,69 @@
+    Supported exit states from Notify dialog.
+    """
+
+    DEFER_UPDATES = "0"
+    CHECK_UPDATES = "1"
```
@gonzalo-bulnes I thought about the random strings/longer ints, but because I am handling QMessageBox's own return values (which can only be `QMessageBox.Ok` or `QMessageBox.No` and returning basically a flag, I think the risk of the wrong thing being returned is minimal.

As for why they are strings, it was for stylistic consistency with the existing enum structure in `Updater.py`, but I am open to changing it and/or changing both at the same time (again to keep consistency).

gonzalo-bulnes replied:
No strong feelings, as you say, they're flags and I too think we'd notice any (unlikely) misuse during review. The type is easy to change anytime if we want to, so I wouldn't overthink it. If we find a strong argument in the future, it will be easy to do whatever makes sense then 👍

rocodes commented:
> Re: caps for constants
> 
> > I'd be happy to switch to more typical Python naming conventions,
> 
> AFAIK the all-caps style is idiomatic, see https://www.python.org/dev/peps/pep-0008/#constants

Yes - I was referring to the filenames themselves (`NotifyApp`, `UpdaterApp` instead of `notify_app`), I should have written camel vs snake case.

creviera reviewed (dismissed):
I can run through the test plan tomorrow. Your code comments are helpful and the new `NotifyDialog` class is a nice abstraction.

rocodes commented on launcher/sdw-notify.py:
```diff
@@ -52,19 +50,34 @@ def main():
+            updater = UpdaterApp.UpdaterApp()
+            updater.show()
+            sys.exit(app.exec_())
+    except ValueError:
```
Yeah, maybe it's a bit clumsy--that was me trying to responsibly handle the scenario where `QMessageDialog.exec_()` returns something we aren't expecting. (Currently the dialog will only return one of `QMessageBox.Yes` or `QMessageBox.No` for all user behaviour, including closing the dialog or using the keyboard keys to manipulate the dialog. But if for example some future changes were made such that the dialog could enter a state where it returns a third option, like a `QMessageBox.Apply, I wanted to be clear that we aren't expecting that kind of behaviour.)
I'll add  comment but I'm also open to rewriting this if you think there's a nicer way to handle it.

rocodes commented:
@creviera Thank you for your review! I've addressed most of your review comments in fca95fb, and have rebased/squashed commits. The only thing I haven't addressed is the caps case in `strings.py`, since `launcher/sdw_update_gui/strings.py` uses lowercase too--I can put in a small formatting PR that addresses both if we decide to standardize on caps for the strings files.

rocodes commented on launcher/sdw_notify/NotifyApp.py:
```diff
@@ -0,0 +1,82 @@
+    CHECK_UPDATES = "0"
+    DEFER_UPDATES = "1"
+    ERROR_UNKNOWN = "2"
+
```
@creviera per your feedback, reordered these and added the error state

creviera commented on launcher/sdw_notify/NotifyApp.py:
```diff
@@ -0,0 +1,82 @@
+    def run(self) -> NotifyStatus:
+        """
+        Launch dialog and return user selection.
+        Raises ValueError if dialog encounters unhandled state.
```
this can be removed now

rocodes replied:
oops. removed in 2c70adf

creviera commented on launcher/sdw_notify/NotifyApp.py:
```diff
@@ -0,0 +1,81 @@
+
+        if result == QMessageBox.Ok:
+            sdlog.info(
+                "NotfyDialog returned {}, user has opted to check for " "updates".format(result)
```
[nit] remove `" "` appended to `updates`

creviera reviewed (approved):
## Testing

Check out this branch in your sd dev vm.

**Manual testing**
Copy this branch to dom0 via `make clone`, then `make dev`. 

Ensure `sd-app` VM is shut down. 
Edit the file in `dom0` called `~/.securedrop_launcher/sdw-last-updated` to change the date to a date >= 5 days ago.
Run `python3 /opt/securedrop/launcher/sdw-notify.py`:
- [x] Notification appears with prompt to check for updates. 
- [x] No "**Warning:**" text is present in the notification
- [x] The default selected button option is "Check for updates".
- [x] Clicking "Remind me later" (or pressing the Esc key) closes the dialog window. Note the time.

Start `sd-app` vm via `qvm-start sd-app` in `dom0`. Wait for the notify prompt to reappear.  
- [ ] The notify prompt appears again after a delay of t < 2 hours (The reminder runs every 2 hours on the even hours. The delay will depend on what time the user dismissed the dialog). 
  * Skipping after confirming that you tested this already
- [x] This time, the "**Warning:**" text is present in the notification
- [x] Clicking "Check for updates" launches the Preflight Updater.  
- [x] The Preflight Updater can run updates successfully. 
- [x] No GTK warnings or other errors are generated or printed to the `dom0` console when the Preflight Updater is launched.

**Automated testing**
- [x] All tests pass

creviera commented:
waiting for updater to finish, but all looks good! skipping the test for the `0 */2 * * *` change as part of the test plan since i confirmed with @rocodes that this was already tested and i don't think we need to wait until noon to confirm it again.

creviera commented:
and updater finished, client is launched, looks lovely :rocket:
